osu.Game.Rulesets.Osu/Skinning/Legacy/LegacyOldStyleSpinner.cs

```diff
@@ -33,39 +33,31 @@ private void load(ISkinSource source)
         {
             spinnerBlink = source.GetConfig<OsuSkinConfiguration, bool>(OsuSkinConfiguration.SpinnerNoBlink)?.Value != true;
 
-            AddInternal(new Container
+            AddRangeInternal(new Drawable[]
             {
-                // the old-style spinner relied heavily on absolute screen-space coordinate values.
-                // wrap everything in a container simulating absolute coords to preserve alignment
-                // as there are skins that depend on it.
-                Width = 640,
-                Height = 480,
-                Anchor = Anchor.Centre,
-                Origin = Anchor.Centre,
-                Children = new Drawable[]
+                new Sprite
                 {
-                    new Sprite
-                    {
-                        Anchor = Anchor.Centre,
-                        Origin = Anchor.Centre,
-                        Texture = source.GetTexture("spinner-background"),
-                        Scale = new Vector2(SPRITE_SCALE)
-                    },
-                    disc = new Sprite
-                    {
-                        Anchor = Anchor.Centre,
-                        Origin = Anchor.Centre,
-                        Texture = source.GetTexture("spinner-circle"),
-                        Scale = new Vector2(SPRITE_SCALE)
-                    },
-                    metre = new Container
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Texture = source.GetTexture("spinner-background"),
+                    Scale = new Vector2(SPRITE_SCALE)
+                },
+                disc = new Sprite
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Texture = source.GetTexture("spinner-circle"),
+                    Scale = new Vector2(SPRITE_SCALE)
+                },
+                new LegacyCoordinatesContainer
```

Comment:
I'm not sure I quite understand this change. From what I'm reading in osu-stable, both `spinner-background` and `spinner-circle` should be included in here, as it was in the original code.

```csharp
protected int spinnerTopOffset = 45;

if (GameBase.GameField.CorrectionOffsetActive)
    spinnerTopOffset -= 16;  // (this is the -8 in our code)

posTopLeftCentre = new Vector2(GameBase.WindowManager.WidthScaled / Math.Max(1, GameBase.WindowManager.WidthWidescreenRatio) / 2f, spinnerTopOffset + 219);

if (newStyleSpinner)
{
    SpriteCircleTop = new pSprite(TextureManager.Load(@"spinner-top"), Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime + 1), false, initialColour);
    SpriteCollection.Add(SpriteCircleTop);

    spriteCircleBottom = new pSprite(TextureManager.Load(@"spinner-bottom"), Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime), false, initialColour);
    SpriteCollection.Add(spriteCircleBottom);

    spriteMiddleTop = new pSprite(TextureManager.Load(@"spinner-middle"), Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime + 3), false, Color.TransparentWhite);
    SpriteCollection.Add(spriteMiddleTop);

    spriteMiddleBottom = new pSprite(TextureManager.Load(@"spinner-middle2"), Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime + 2), false, Color.TransparentWhite);
    SpriteCollection.Add(spriteMiddleBottom);
}
else
{
    pTexture bgTexture = null;
    //Check for a jpg background for beatmap-based skins (used to reduce filesize), then fallback to png.
    if (!SkinManager.IgnoreBeatmapSkin)
        bgTexture = TextureManager.Load(@"spinner-background.jpg", SkinSource.Beatmap);
    if (bgTexture == null)
        bgTexture = TextureManager.Load(@"spinner-background", SkinSource.Beatmap | SkinSource.Skin);

    Color bgColour = SkinManager.Colours.ContainsKey(@"SpinnerBackground") ? SkinManager.Colours[@"SpinnerBackground"] : new Color(100, 100, 100, 255);
    spriteBackground = new pSprite(bgTexture, Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime - 1), false, bgColour);
    SpriteCollection.Add(spriteBackground);

    SpriteCircleTop = new pSprite(TextureManager.Load(@"spinner-circle"), Fields.TopLeft, Origins.Centre, Clocks.Audio, posTopLeftCentre, SpriteManager.drawOrderFwdLowPrio(StartTime), false, initialColour);
    SpriteCollection.Add(SpriteCircleTop);

    spriteScoreMetre = new pSprite(TextureManager.Load(@"spinner-metre"), Fields.TopLeft, Origins.TopLeft, Clocks.Audio, new Vector2(posTopLeftCentre.X - 320, spinnerTopOffset), SpriteManager.drawOrderFwdLowPrio(StartTime + 1), false, initialColour);
    spriteScoreMetre.DrawHeight = 0;
    SpriteCollection.Add(spriteScoreMetre);
}
```

What am I missing?

Reply:
I've not put them in the `LegacyCoordinatesContainer` to make centering them as simple as just setting anchor and origin to centre, otherwise I would need to declare a public constant just like osu!stable (`SPINNER_CENTRE_Y`):
```diff
- Anchor = Anchor.Centre,
- Origin = Anchor.Centre,
+ Anchor = Anchor.TopCentre,
+ Origin = Anchor.Centre,
+ Y = SPINNER_CENTRE_Y, // spinner_top_offset + 219f
```

I'm fine with doing it either way, and would actually remove the necessity of exposing `LegacyCoordinatesContainer` if including bonus counter and spm counter inside `LegacySpinner` is deemed the way here.

Reply:
Well the part that confuses me is all of them are `posTopLeftCentre` except for the metre which is just `spinnerTopOffset`. But `posTopLeftCentre` includes the -16 (/-8) offset in its y-coord, whereas here it looks to be removed and only factored in for the metre?

Reply:
I don't follow? `spinnerTopOffset + 219` is the equivalent of `Anchor = Origin = Anchor.Centre`, what do you mean by removed?

Reply:
You keep saying that it's equivalent to centering without the -8 adjustment but how have you determined this? Only visually? Honestly the current structure is too confusing for me to confirm that with actual math.

Reply:
I've had to go in via manual math to confirm that
`posTopLeftCentre = spinnerTopOffset + 219`
`= 45 - 16 + 219`
`= 480 / 2 + 8`
(i.e. half the window size + 8, so centred in the gamefield while in the window coordinate space since `Fields.TopLeft`).

But yes it's very confusing...

Reply:
Right, I see. I'd still rather do the const thing to match source 1:1.